src/util/recentlyEditedTree.js

```diff
@@ -0,0 +1,307 @@
+      For example:
+
+      Recently edited list
+
+      A.B.T.U
+      A.B.C.D.E.F
+      A.B.C.G.H.I.J
+      A.B.C.G.H.K.L
+
+      If A.B.C.D.E is the deleted path then
+      closestAncestor with multiple children is A.B.C
+      diff between deletedNode (A.B.C.D.E) and closestAncestor (A.B.C) is 2
+      Now pathBeingMerged is A.B.C.D and we delete it.
+      Then we update all the descendants of closestAncestor i.e A.B.C.G.H.I.J and A.B.C.G.H.K.L.
+      */
+      const pathBeingMerged = commonPath.slice(0, closestAncestor.path.length + 1)
+      _.unset(tree, pathBeingMerged) // deleting the merged path
+      findTreeDescendants(tree, nodeToMergeIntoPath).forEach(descendant => {
+        _.set(tree, pathToContext(descendant.path), { leaf: true, lastUpdated: timestampUpdate ? timestamp() : descendant.lastUpdated, path: descendant.path })
+      })
+    }
+    /*
+     if cannot be merged to either parent or grandParent then just making parent path as the leaf node
+
+      For example:
+      Recently edited list
+
+      A.B.T.U
+      A.B.C.D.E.F
+      A.B.C.G.H.I.J
+      A.B.C.G.H.K.L
+
+      If A.B.C.D.E.F is deleted path
+      closestAncestor with multiple children is A.B.C
+      Diff between deletedNode (A.B.C.D.E.F) and closestAncestor (A.B.C) is 3.
+      So no need to merge into other branches , we just delete A.B.C.D.E.F and set A.B.D.E as leaf node and update it.
+     */
+    else {
+      _.unset(tree, commonPath)
+      _.set(tree, contextOf(commonPath), { leaf: true, lastUpdated: timestamp(), path: oldPath.slice(0, contextOf(commonPath).length) })
+    }
+  }
+}
+
+/**
+   * Moves a node between branches in the existing tree object using mutation.
+   * @param {object} tree nested object representing tree
+   * @param {string[]} oldPath array of string representing path
+   * @param {string[]} newPath array of string representing path
+   */
+const nodeMove = (tree, oldPath, newPath) => {
+  const oldContext = pathToContext(oldPath)
+  const newContext = pathToContext(newPath)
+
+  const { node: oldNode, path: commonOldPath } = findTreeDeepestSubcontext(tree, oldContext)
+  const { path: commonNewPath } = findTreeDeepestSubcontext(tree, newContext)
+
+  if (oldNode && commonOldPath.length === oldContext.length) {
+    // if node is already available
+    if (equalArrays(commonOldPath, commonNewPath)) {
+      // if moved with in same place (only rank has changed) (to prevent traversing descendants)
+      nodeChange(tree, oldPath, newPath)
+    }
+    else if (oldNode.leaf) {
+      nodeAdd(tree, contextOf(newPath), [head(newPath)])
+      nodeDelete(tree, oldPath, false)
+    }
+    else {
+      const descendants = findTreeDescendants(tree, oldContext)
+      descendants.forEach(descendant => {
+        const addedPath = [head(newPath)].concat(descendant.path.slice(oldPath.length))
+        nodeAdd(tree, contextOf(newPath), addedPath)
```

Comment:
`nodeAdd(newPath.concat(descendant.path.slice(oldPath.length))`

*This differs slightly from the current implementation. I think it will still work but correct me if I am wrong. I am hoping that `nodeAdd` can be used the same way in each conditional branch, albeit with different arguments, so that it does not have to have a special flag for this case.

Reply:
About this I just realized that with the current implementation of `nodeChange` , we can add new node by passing `newPath` as both `oldPath` and `newPath` parameter of `nodeChange`. It is because `commonPath` will be be derived from `oldPath` anyways. So this change will eliminate the need of `DUMMY_TOKEN` too.

Reply:
Aha! My suspicion was correct :). I had assumed `oldPath` was derivable from `newPath`, but I didn't know it would be as easy as passing `newPath` as `oldPath` given the way that `commonPath` is found. Looks great.